## IMDB Movie Business Use-case


- The database contains info of several years about:
    - Movies
    - Rating
    - Director
    - Popularity
    - Revenue & Budget



In [2]:
import numpy as np
import pandas as pd
import os

In [3]:
os.getcwd()

'c:\\Users\\pmgho\\Downloads\\Prajwal Machine Learning Data Set\\python\\List Question\\pandas with prajwal'

In [4]:
os.chdir('C:\\Users\\pmgho\\Downloads\\Prajwal Machine Learning Data Set\\')

In [5]:
os.getcwd()

'C:\\Users\\pmgho\\Downloads\\Prajwal Machine Learning Data Set'

## Reading the dataset

In [6]:
movies = pd.read_csv('movies.csv')
movies

Unnamed: 0     id     budget  popularity     revenue  \
0              0  43597  237000000         150  2787965087   
1              1  43598  300000000         139   961000000   
2              2  43599  245000000         107   880674609   
3              3  43600  250000000         112  1084939099   
4              5  43602  258000000         115   890871626   
...          ...    ...        ...         ...         ...   
1460        4736  48363          0           3      321952   
1461        4743  48370      27000          19     3151130   
1462        4748  48375          0           7           0   
1463        4749  48376          0           3           0   
1464        4768  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day  
0            4762  2009   Dec  Thursday  
1            4763  2007   May  Saturday  
2            4764  2015   Oct    Monday  
3            4765  2012   Jul    Monday  
4            4767  2007   May   Tuesday  
...           ...   ...   ...       ...  
1460         4809  1978   May    Monday  
1461         5369  1994   Sep   Tuesday  
1462         5148  2009   Aug    Friday  
1463         5535  1990   Jul    Friday  
1464         5097  1992   Sep    Friday  

[1465 rows x 12 columns]

In [7]:
movies.head()

Unnamed: 0     id     budget  popularity     revenue  \
0           0  43597  237000000         150  2787965087   
1           1  43598  300000000         139   961000000   
2           2  43599  245000000         107   880674609   
3           3  43600  250000000         112  1084939099   
4           5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
4         4767  2007   May   Tuesday


#### what kind of questions can we ask from this dataset?

   - **Top 10 most popular movies**
   - **highest rated movies**
   - **number of movies released per year**
   - **find highest budget movies in a year**

#### Can we ask more interesting/deeper questions?
   - Find the most productive director?
   - Which director produces high budget films?
   - Highest and lowest rated movies for every month in a particular year.



Notice, that we also get a column **Unnamed: 0** which represents nothing but the index of a row.

We can simply add one more argument `index_col=0`(treat first column as index) to get rid of this

The default value is `index_col=None`.



In [8]:
movies = pd.read_csv('movies.csv',index_col=0)
movies

id     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
5     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
4736  48363          0           3      321952   
4743  48370      27000          19     3151130   
4748  48375          0           7           0   
4749  48376          0           3           0   
4768  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
5                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
4736                            The Last Waltz           7.9          64   
4743                                    Clerks           7.4         755   
4748                                   Rampage           6.0         131   
4749                                   Slacker           6.4          77   
4768                               El Mariachi           6.6         238   

      director_id  year month       day  
0            4762  2009   Dec  Thursday  
1            4763  2007   May  Saturday  
2            4764  2015   Oct    Monday  
3            4765  2012   Jul    Monday  
5            4767  2007   May   Tuesday  
...           ...   ...   ...       ...  
4736         4809  1978   May    Monday  
4743         5369  1994   Sep   Tuesday  
4748         5148  2009   Aug    Friday  
4749         5535  1990   Jul    Friday  
4768         5097  1992   Sep    Friday  

[1465 rows x 11 columns]

In [9]:
movies.head()

id     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
5  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
5                              Spider-Man 3           5.9        3576   

   director_id  year month       day  
0         4762  2009   Dec  Thursday  
1         4763  2007   May  Saturday  
2         4764  2015   Oct    Monday  
3         4765  2012   Jul    Monday  
5         4767  2007   May   Tuesday

In [10]:
#Lets check the shape of dataset:
movies.shape

(1465, 11)

In [11]:
directors = pd.read_csv('directors.csv')
directors

Unnamed: 0      director_name    id gender
0              0      James Cameron  4762   Male
1              1     Gore Verbinski  4763   Male
2              2         Sam Mendes  4764   Male
3              3  Christopher Nolan  4765   Male
4              4     Andrew Stanton  4766   Male
...          ...                ...   ...    ...
2344        2344      Shane Carruth  7106   Male
2345        2345   Neill Dela Llana  7107    NaN
2346        2346        Scott Smith  7108    NaN
2347        2347        Daniel Hsia  7109   Male
2348        2348   Brian Herzlinger  7110   Male

[2349 rows x 4 columns]

In [12]:
directors = pd.read_csv('directors.csv',index_col=0)
directors

director_name    id gender
0         James Cameron  4762   Male
1        Gore Verbinski  4763   Male
2            Sam Mendes  4764   Male
3     Christopher Nolan  4765   Male
4        Andrew Stanton  4766   Male
...                 ...   ...    ...
2344      Shane Carruth  7106   Male
2345   Neill Dela Llana  7107    NaN
2346        Scott Smith  7108    NaN
2347        Daniel Hsia  7109   Male
2348   Brian Herzlinger  7110   Male

[2349 rows x 3 columns]

In [13]:
directors.head()

director_name    id gender
0      James Cameron  4762   Male
1     Gore Verbinski  4763   Male
2         Sam Mendes  4764   Male
3  Christopher Nolan  4765   Male
4     Andrew Stanton  4766   Male

In [14]:
directors.shape

(2349, 3)

### Merging of both Dataframe:

#### So we want to include directors df info into movies df. How can we do this ?

  - `merge()`


But, before merging, check if for all the movies in `movies` df, have their corresponding director details present in the `directors` df or not.





#### How do we get the number of unique directors in `movies`?

We can do it using `nunique()`

- **`unique()`** gives **unique values** in a column

- **`nunique()`** gives **number of unique values** in a column



In [15]:
movies['director_id'].nunique()

199

In [16]:
directors['id'].nunique()

2349

- Movies Dataset: 1465 rows, but only 199 unique directors
- Directors Dataset: 2349 unique directors (= no of rows)


#### Inference?

- directors in `movies` is a subset of directors in `directors`
- need to check if we have details for 199 directors present in `directors` df also

#### How to check whether all the values in director_id column of movies is present in id column of director ?

- `isin()` method

#### Cant we do this using Python `in`?

- We can, but this will work for one element at a time.
- We need to do this for all the values in the column
- The `isin()` method checks if the Dataframe column contains the specified value(s).



  

In [17]:
movies['director_id'].isin(directors['id'])

0       True
1       True
2       True
3       True
5       True
        ... 
4736    True
4743    True
4748    True
4749    True
4768    True
Name: director_id, Length: 1465, dtype: bool

In [18]:
np.all(movies['director_id'].isin(directors['id']))

np.True_

In [19]:
np.all(movies['director_id'].isin(directors['id']))

np.True_

We can use  `LEFT OUTER JOIN` to merge

#### Recall what will Left Outer Join do?
Left outer join will include all the rows of df `movies` and only those from `directors` that match with values of `movies`['director_id']

<img src="https://www.datasciencemadesimple.com/wp-content/uploads/2017/09/join-or-merge-in-python-pandas-1.png?ezimgfmt=ng:webp/ngcb1" />





In [20]:
# if column name is not same
# `left_on`: Specifies the key of the 1st dataframe
# `right_on`: Specifies the key of the 2nd dataframe

data = movies.merge(directors,how='left',left_on='director_id',right_on='id')
data

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name  id_y gender  
0            4762  2009   Dec  Thursday      James Cameron  4762   Male  
1            4763  2007   May  Saturday     Gore Verbinski  4763   Male  
2            4764  2015   Oct    Monday         Sam Mendes  4764   Male  
3            4765  2012   Jul    Monday  Christopher Nolan  4765   Male  
4            4767  2007   May   Tuesday          Sam Raimi  4767   Male  
...           ...   ...   ...       ...                ...   ...    ...  
1460         4809  1978   May    Monday    Martin Scorsese  4809   Male  
1461         5369  1994   Sep   Tuesday        Kevin Smith  5369   Male  
1462         5148  2009   Aug    Friday           Uwe Boll  5148   Male  
1463         5535  1990   Jul    Friday  Richard Linklater  5535   Male  
1464         5097  1992   Sep    Friday   Robert Rodriguez  5097    NaN  

[1465 rows x 14 columns]

Since the columns with name `id`is present in both the df.

After merging:

- id_x : represents id values from movie df
- id_y: represents id values from directors df



### Dropping redundant columns

In [21]:
data.drop(['id_y'],axis=1,inplace=True)
data.head()

id_x     budget  popularity     revenue  \
0  43597  237000000         150  2787965087   
1  43598  300000000         139   961000000   
2  43599  245000000         107   880674609   
3  43600  250000000         112  1084939099   
4  43602  258000000         115   890871626   

                                      title  vote_average  vote_count  \
0                                    Avatar           7.2       11800   
1  Pirates of the Caribbean: At World's End           6.9        4500   
2                                   Spectre           6.3        4466   
3                     The Dark Knight Rises           7.6        9106   
4                              Spider-Man 3           5.9        3576   

   director_id  year month       day      director_name gender  
0         4762  2009   Dec  Thursday      James Cameron   Male  
1         4763  2007   May  Saturday     Gore Verbinski   Male  
2         4764  2015   Oct    Monday         Sam Mendes   Male  
3         4765  2012   Jul    Monday  Christopher Nolan   Male  
4         4767  2007   May   Tuesday          Sam Raimi   Male

## Feature Exploration





In [22]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id_x           1465 non-null   int64  
 1   budget         1465 non-null   int64  
 2   popularity     1465 non-null   int64  
 3   revenue        1465 non-null   int64  
 4   title          1465 non-null   object 
 5   vote_average   1465 non-null   float64
 6   vote_count     1465 non-null   int64  
 7   director_id    1465 non-null   int64  
 8   year           1465 non-null   int64  
 9   month          1465 non-null   object 
 10  day            1465 non-null   object 
 11  director_name  1465 non-null   object 
 12  gender         1341 non-null   object 
dtypes: float64(1), int64(7), object(5)
memory usage: 148.9+ KB


In [23]:
def encode(data):
    if data == 'Male':
        return 0
    else:
        return 1

In [30]:
data['gender'] = data['gender'].apply(encode)
data

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name  gender  
0            4762  2009   Dec  Thursday      James Cameron       0  
1            4763  2007   May  Saturday     Gore Verbinski       0  
2            4764  2015   Oct    Monday         Sam Mendes       0  
3            4765  2012   Jul    Monday  Christopher Nolan       0  
4            4767  2007   May   Tuesday          Sam Raimi       0  
...           ...   ...   ...       ...                ...     ...  
1460         4809  1978   May    Monday    Martin Scorsese       0  
1461         5369  1994   Sep   Tuesday        Kevin Smith       0  
1462         5148  2009   Aug    Friday           Uwe Boll       0  
1463         5535  1990   Jul    Friday  Richard Linklater       0  
1464         5097  1992   Sep    Friday   Robert Rodriguez       1  

[1465 rows x 13 columns]

In [24]:
data['gender'].value_counts()

gender
Male      1309
Female      32
Name: count, dtype: int64

In [25]:
data[['revenue','budget']].apply(np.sum)

revenue    209866997305
budget      70353617179
dtype: int64

In [26]:
data[['revenue','budget']].apply(np.sum,axis=1)

0       3024965087
1       1261000000
2       1125674609
3       1334939099
4       1148871626
           ...    
1460        321952
1461       3178130
1462             0
1463             0
1464       2260920
Length: 1465, dtype: int64

In [27]:
# we define a funcion to calculate profit

def profit(x):
    return x['revenue']-x['budget']
data['profit'] = data[['revenue','budget']].apply(profit,axis=1)
data

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name gender      profit  
0            4762  2009   Dec  Thursday      James Cameron   Male  2550965087  
1            4763  2007   May  Saturday     Gore Verbinski   Male   661000000  
2            4764  2015   Oct    Monday         Sam Mendes   Male   635674609  
3            4765  2012   Jul    Monday  Christopher Nolan   Male   834939099  
4            4767  2007   May   Tuesday          Sam Raimi   Male   632871626  
...           ...   ...   ...       ...                ...    ...         ...  
1460         4809  1978   May    Monday    Martin Scorsese   Male      321952  
1461         5369  1994   Sep   Tuesday        Kevin Smith   Male     3124130  
1462         5148  2009   Aug    Friday           Uwe Boll   Male           0  
1463         5535  1990   Jul    Friday  Richard Linklater   Male           0  
1464         5097  1992   Sep    Friday   Robert Rodriguez    NaN     1820920  

[1465 rows x 14 columns]

In [28]:
data.groupby('director_name')

In [29]:
data.groupby('director_name').get_group('Alexander Payne')

id_x    budget  popularity    revenue            title  vote_average  \
793   45163  30000000          19  105834556    About Schmidt           6.7   
1006  45699  20000000          40  177243185  The Descendants           6.7   
1101  46004  16000000          23  109502303         Sideways           6.9   
1211  46446  12000000          29   17654912         Nebraska           7.4   
1281  46813         0          13          0         Election           6.7   

      vote_count  director_id  year month       day    director_name gender  \
793          362         5451  2002   Dec    Friday  Alexander Payne    NaN   
1006         934         5451  2011   Sep    Friday  Alexander Payne    NaN   
1101         478         5451  2004   Oct    Friday  Alexander Payne    NaN   
1211         636         5451  2013   Sep  Saturday  Alexander Payne    NaN   
1281         270         5451  1999   Apr    Friday  Alexander Payne    NaN   

         profit  
793    75834556  
1006  157243185  
1101   93502303  
1211    5654912  
1281          0

In [30]:
data.groupby('director_name')['title'].count()

director_name
Adam McKay                      6
Adam Shankman                   8
Alejandro González Iñárritu     6
Alex Proyas                     5
Alexander Payne                 5
                               ..
Wes Craven                     10
Wolfgang Petersen               7
Woody Allen                    18
Zack Snyder                     7
Zhang Yimou                     6
Name: title, Length: 199, dtype: int64

In [31]:
data

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name gender      profit  
0            4762  2009   Dec  Thursday      James Cameron   Male  2550965087  
1            4763  2007   May  Saturday     Gore Verbinski   Male   661000000  
2            4764  2015   Oct    Monday         Sam Mendes   Male   635674609  
3            4765  2012   Jul    Monday  Christopher Nolan   Male   834939099  
4            4767  2007   May   Tuesday          Sam Raimi   Male   632871626  
...           ...   ...   ...       ...                ...    ...         ...  
1460         4809  1978   May    Monday    Martin Scorsese   Male      321952  
1461         5369  1994   Sep   Tuesday        Kevin Smith   Male     3124130  
1462         5148  2009   Aug    Friday           Uwe Boll   Male           0  
1463         5535  1990   Jul    Friday  Richard Linklater   Male           0  
1464         5097  1992   Sep    Friday   Robert Rodriguez    NaN     1820920  

[1465 rows x 14 columns]

In [32]:
data.describe()

id_x        budget   popularity       revenue  vote_average  \
count   1465.000000  1.465000e+03  1465.000000  1.465000e+03   1465.000000   
mean   45225.191126  4.802295e+07    30.855973  1.432539e+08      6.368191   
std     1189.096396  4.935541e+07    34.845214  2.064918e+08      0.818033   
min    43597.000000  0.000000e+00     0.000000  0.000000e+00      3.000000   
25%    44236.000000  1.400000e+07    11.000000  1.738013e+07      5.900000   
50%    45022.000000  3.300000e+07    23.000000  7.578164e+07      6.400000   
75%    45990.000000  6.600000e+07    41.000000  1.792469e+08      6.900000   
max    48395.000000  3.800000e+08   724.000000  2.787965e+09      8.300000   

         vote_count  director_id         year        profit  
count   1465.000000  1465.000000  1465.000000  1.465000e+03  
mean    1146.396587  5040.192491  2002.615017  9.523098e+07  
std     1578.077438   258.059631     8.680141  1.751670e+08  
min        1.000000  4762.000000  1976.000000 -1.657101e+08  
25%      216.000000  4845.000000  1998.000000  0.000000e+00  
50%      571.000000  4964.000000  2004.000000  3.797725e+07  
75%     1387.000000  5179.000000  2009.000000  1.268836e+08  
max    13752.000000  6204.000000  2016.000000  2.550965e+09

In [33]:
data.describe(include=object)

title month     day     director_name gender
count          1465  1465    1465              1465   1341
unique         1465    12       7               199      2
top     El Mariachi   Dec  Friday  Steven Spielberg   Male
freq              1   193     654                26   1309

### How to find multiple aggregates of any feature?

#### Finding the very first year and the latest year a director released a movie i.e. basically the min of the ***year*** column,grouped by ***director_name***.

In [34]:
data.groupby(['director_name'])['year'].aggregate(['min','max'])

min   max
director_name                          
Adam McKay                   2004  2015
Adam Shankman                2001  2012
Alejandro González Iñárritu  2000  2015
Alex Proyas                  1994  2016
Alexander Payne              1999  2013
...                           ...   ...
Wes Craven                   1984  2011
Wolfgang Petersen            1981  2006
Woody Allen                  1977  2013
Zack Snyder                  2004  2016
Zhang Yimou                  2002  2014

[199 rows x 2 columns]

### Group based Filtering

**> Group based filtering allows us to filter rows from each group by using conditional statements on data each group rather
than the whole dataframe.**

***How to find the details of the movies by high budget directors?***

***> Lets assume, high budget director-> any director with atleast one movie with budget>100M***

#### 1) We can group the data by director and use ***max*** of the budget as aggregator.

In [35]:
data_dir_budget=data.groupby('director_name')['budget'].max().reset_index()
data_dir_budget.head()

director_name     budget
0                   Adam McKay  100000000
1                Adam Shankman   80000000
2  Alejandro González Iñárritu  135000000
3                  Alex Proyas  140000000
4              Alexander Payne   30000000

### We can the **filter** out the director names with max budget>100M

In [39]:
names = data_dir_budget.loc[data_dir_budget['budget'] >= 100,'director_name']
display(names)

0                       Adam McKay
1                    Adam Shankman
2      Alejandro González Iñárritu
3                      Alex Proyas
4                  Alexander Payne
                  ...             
194                     Wes Craven
195              Wolfgang Petersen
196                    Woody Allen
197                    Zack Snyder
198                    Zhang Yimou
Name: director_name, Length: 199, dtype: object

In [38]:
names.shape

(199,)

In [41]:
data.loc[data["director_name"].isin(names)]

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name gender      profit  
0            4762  2009   Dec  Thursday      James Cameron   Male  2550965087  
1            4763  2007   May  Saturday     Gore Verbinski   Male   661000000  
2            4764  2015   Oct    Monday         Sam Mendes   Male   635674609  
3            4765  2012   Jul    Monday  Christopher Nolan   Male   834939099  
4            4767  2007   May   Tuesday          Sam Raimi   Male   632871626  
...           ...   ...   ...       ...                ...    ...         ...  
1460         4809  1978   May    Monday    Martin Scorsese   Male      321952  
1461         5369  1994   Sep   Tuesday        Kevin Smith   Male     3124130  
1462         5148  2009   Aug    Friday           Uwe Boll   Male           0  
1463         5535  1990   Jul    Friday  Richard Linklater   Male           0  
1464         5097  1992   Sep    Friday   Robert Rodriguez    NaN     1820920  

[1465 rows x 14 columns]

#### Can we filter groups in a single go using Lambda functions? Yes!

In [42]:
data.groupby('director_name').filter(lambda x: x['budget'].max() >= 100)

id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name gender      profit  
0            4762  2009   Dec  Thursday      James Cameron   Male  2550965087  
1            4763  2007   May  Saturday     Gore Verbinski   Male   661000000  
2            4764  2015   Oct    Monday         Sam Mendes   Male   635674609  
3            4765  2012   Jul    Monday  Christopher Nolan   Male   834939099  
4            4767  2007   May   Tuesday          Sam Raimi   Male   632871626  
...           ...   ...   ...       ...                ...    ...         ...  
1460         4809  1978   May    Monday    Martin Scorsese   Male      321952  
1461         5369  1994   Sep   Tuesday        Kevin Smith   Male     3124130  
1462         5148  2009   Aug    Friday           Uwe Boll   Male           0  
1463         5535  1990   Jul    Friday  Richard Linklater   Male           0  
1464         5097  1992   Sep    Friday   Robert Rodriguez    NaN     1820920  

[1465 rows x 14 columns]

#### Group based Apply

### * Applying a function on grouped objects

**>What if we to do the transformation of a column using some column's agreegate**

***Let's say we want to filter the risky movies whose budget was even higher than the revenue of the director from his movies.***

***We can subtract the average *revenue* of a director from *budget* column, for each director.***

In [45]:
def func(x):
    # return whether a movie is risky or not.
    x['risky'] = x['budget'] - x['revenue'].mean() >= 0
    return x

data_risky = data.groupby('director_name',group_keys=False).apply(func)
data_risky

C:\Users\pmgho\AppData\Local\Temp\ipykernel_4656\2816659029.py:6: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  data_risky = data.groupby('director_name',group_keys=False).apply(func)


id_x     budget  popularity     revenue  \
0     43597  237000000         150  2787965087   
1     43598  300000000         139   961000000   
2     43599  245000000         107   880674609   
3     43600  250000000         112  1084939099   
4     43602  258000000         115   890871626   
...     ...        ...         ...         ...   
1460  48363          0           3      321952   
1461  48370      27000          19     3151130   
1462  48375          0           7           0   
1463  48376          0           3           0   
1464  48395     220000          14     2040920   

                                         title  vote_average  vote_count  \
0                                       Avatar           7.2       11800   
1     Pirates of the Caribbean: At World's End           6.9        4500   
2                                      Spectre           6.3        4466   
3                        The Dark Knight Rises           7.6        9106   
4                                 Spider-Man 3           5.9        3576   
...                                        ...           ...         ...   
1460                            The Last Waltz           7.9          64   
1461                                    Clerks           7.4         755   
1462                                   Rampage           6.0         131   
1463                                   Slacker           6.4          77   
1464                               El Mariachi           6.6         238   

      director_id  year month       day      director_name gender      profit  \
0            4762  2009   Dec  Thursday      James Cameron   Male  2550965087   
1            4763  2007   May  Saturday     Gore Verbinski   Male   661000000   
2            4764  2015   Oct    Monday         Sam Mendes   Male   635674609   
3            4765  2012   Jul    Monday  Christopher Nolan   Male   834939099   
4            4767  2007   May   Tuesday          Sam Raimi   Male   632871626   
...           ...   ...   ...       ...                ...    ...         ...   
1460         4809  1978   May    Monday    Martin Scorsese   Male      321952   
1461         5369  1994   Sep   Tuesday        Kevin Smith   Male     3124130   
1462         5148  2009   Aug    Friday           Uwe Boll   Male           0   
1463         5535  1990   Jul    Friday  Richard Linklater   Male           0   
1464         5097  1992   Sep    Friday   Robert Rodriguez    NaN     1820920   

      risky  
0     False  
1     False  
2     False  
3     False  
4     False  
...     ...  
1460  False  
1461  False  
1462  False  
1463  False  
1464  False  

[1465 rows x 15 columns]

In [46]:
data_risky.loc[data_risky['risky']]

id_x     budget  popularity     revenue  \
7     43608  200000000         107   586090727   
12    43614  380000000         135  1045713802   
15    43618  200000000          37   310669540   
20    43624  209000000          64   303025485   
24    43630  210000000           3   459359555   
...     ...        ...         ...         ...   
1347  47224    5000000           7     3263585   
1349  47229    5000000           3     4842699   
1351  47233    5000000           6           0   
1356  47263   15000000          10           0   
1383  47453    3500000           4           0   

                                            title  vote_average  vote_count  \
7                               Quantum of Solace           6.1        2965   
12    Pirates of the Caribbean: On Stranger Tides           6.4        4948   
15                                     Robin Hood           6.2        1398   
20                                     Battleship           5.5        2114   
24                          X-Men: The Last Stand           6.3        3525   
...                                           ...           ...         ...   
1347                          The Sweet Hereafter           6.8         103   
1349                         90 Minutes in Heaven           5.4          40   
1351                                Light Sleeper           5.7          15   
1356                           Dying of the Light           4.5         118   
1383                  In the Name of the King III           3.3          19   

      director_id  year month        day   director_name gender     profit  \
7            4773  2008   Oct   Thursday    Marc Forster   Male  386090727   
12           4775  2011   May   Saturday    Rob Marshall   Male  665713802   
15           4779  2010   May  Wednesday    Ridley Scott   Male  110669540   
20           4782  2012   Apr  Wednesday      Peter Berg   Male   94025485   
24           4786  2006   May  Wednesday    Brett Ratner   Male  249359555   
...           ...   ...   ...        ...             ...    ...        ...   
1347         5599  1997   May  Wednesday     Atom Egoyan   Male   -1736415   
1349         5951  2015   Sep     Friday  Michael Polish   Male    -157301   
1351         5970  1992   Aug     Friday   Paul Schrader    NaN   -5000000   
1356         5970  2014   Dec   Thursday   Paul Schrader    NaN  -15000000   
1383         5148  2013   Dec     Friday        Uwe Boll   Male   -3500000   

      risky  
7      True  
12     True  
15     True  
20     True  
24     True  
...     ...  
1347   True  
1349   True  
1351   True  
1356   True  
1383   True  

[131 rows x 15 columns]

In [47]:
data_risky.loc[data_risky['risky']==True]

id_x     budget  popularity     revenue  \
7     43608  200000000         107   586090727   
12    43614  380000000         135  1045713802   
15    43618  200000000          37   310669540   
20    43624  209000000          64   303025485   
24    43630  210000000           3   459359555   
...     ...        ...         ...         ...   
1347  47224    5000000           7     3263585   
1349  47229    5000000           3     4842699   
1351  47233    5000000           6           0   
1356  47263   15000000          10           0   
1383  47453    3500000           4           0   

                                            title  vote_average  vote_count  \
7                               Quantum of Solace           6.1        2965   
12    Pirates of the Caribbean: On Stranger Tides           6.4        4948   
15                                     Robin Hood           6.2        1398   
20                                     Battleship           5.5        2114   
24                          X-Men: The Last Stand           6.3        3525   
...                                           ...           ...         ...   
1347                          The Sweet Hereafter           6.8         103   
1349                         90 Minutes in Heaven           5.4          40   
1351                                Light Sleeper           5.7          15   
1356                           Dying of the Light           4.5         118   
1383                  In the Name of the King III           3.3          19   

      director_id  year month        day   director_name gender     profit  \
7            4773  2008   Oct   Thursday    Marc Forster   Male  386090727   
12           4775  2011   May   Saturday    Rob Marshall   Male  665713802   
15           4779  2010   May  Wednesday    Ridley Scott   Male  110669540   
20           4782  2012   Apr  Wednesday      Peter Berg   Male   94025485   
24           4786  2006   May  Wednesday    Brett Ratner   Male  249359555   
...           ...   ...   ...        ...             ...    ...        ...   
1347         5599  1997   May  Wednesday     Atom Egoyan   Male   -1736415   
1349         5951  2015   Sep     Friday  Michael Polish   Male    -157301   
1351         5970  1992   Aug     Friday   Paul Schrader    NaN   -5000000   
1356         5970  2014   Dec   Thursday   Paul Schrader    NaN  -15000000   
1383         5148  2013   Dec     Friday        Uwe Boll   Male   -3500000   

      risky  
7      True  
12     True  
15     True  
20     True  
24     True  
...     ...  
1347   True  
1349   True  
1351   True  
1356   True  
1383   True  

[131 rows x 15 columns]

In [48]:
data.groupby(['director_name'])['title'].count().sort_values(ascending=False)

director_name
Steven Spielberg    26
Martin Scorsese     19
Clint Eastwood      19
Woody Allen         18
Robert Rodriguez    16
                    ..
Stephen Daldry       5
Tom Tykwer           5
Tim Hill             5
Uwe Boll             5
Wayne Wang           5
Name: title, Length: 199, dtype: int64

### Steven Spielberg has directed maximum number of movies

***> But does it make Steven the most productive directors?***

**-->Chances are,he might be active for more years than the other directors.**

***> Calculating the active years for every directors?***

**-->We can substract both min and max year.**

In [69]:
data_agg = data.groupby(['director_name'])[['year','title']].aggregate({'year':['min','max'],'title':'count'})
data_agg

year       title
                              min   max count
director_name                                
Adam McKay                   2004  2015     6
Adam Shankman                2001  2012     8
Alejandro González Iñárritu  2000  2015     6
Alex Proyas                  1994  2016     5
Alexander Payne              1999  2013     5
...                           ...   ...   ...
Wes Craven                   1984  2011    10
Wolfgang Petersen            1981  2006     7
Woody Allen                  1977  2013    18
Zack Snyder                  2004  2016     7
Zhang Yimou                  2002  2014     6

[199 rows x 3 columns]

### Using the new features, can we find the most productive director?

**1) First calculate how many years the director has been active.**

In [70]:
print(data_agg.columns)


MultiIndex([( 'year',   'min'),
            ( 'year',   'max'),
            ('title', 'count')],
           )


In [71]:
data_agg.columns=['_'.join(col) for col in data_agg.columns]
data_agg

year_min  year_max  title_count
director_name                                               
Adam McKay                       2004      2015            6
Adam Shankman                    2001      2012            8
Alejandro González Iñárritu      2000      2015            6
Alex Proyas                      1994      2016            5
Alexander Payne                  1999      2013            5
...                               ...       ...          ...
Wes Craven                       1984      2011           10
Wolfgang Petersen                1981      2006            7
Woody Allen                      1977      2013           18
Zack Snyder                      2004      2016            7
Zhang Yimou                      2002      2014            6

[199 rows x 3 columns]

### Since these were tuples,we can just join them.

#### The columns look good, but we may want to turn back the row labels into a proper column as well.

***Converting row labels into a column using reset_index***

In [73]:
data_agg.reset_index()

director_name  year_min  year_max  title_count
0                     Adam McKay      2004      2015            6
1                  Adam Shankman      2001      2012            8
2    Alejandro González Iñárritu      2000      2015            6
3                    Alex Proyas      1994      2016            5
4                Alexander Payne      1999      2013            5
..                           ...       ...       ...          ...
194                   Wes Craven      1984      2011           10
195            Wolfgang Petersen      1981      2006            7
196                  Woody Allen      1977      2013           18
197                  Zack Snyder      2004      2016            7
198                  Zhang Yimou      2002      2014            6

[199 rows x 4 columns]

### Using the new features, can we find the most productive directors?

***1) First calculate how many years the directors has been active.***

In [74]:
data_agg['yrs_active'] = data_agg['year_max'] - data_agg['year_min']
data_agg

year_min  year_max  title_count  yrs_active
director_name                                                           
Adam McKay                       2004      2015            6          11
Adam Shankman                    2001      2012            8          11
Alejandro González Iñárritu      2000      2015            6          15
Alex Proyas                      1994      2016            5          22
Alexander Payne                  1999      2013            5          14
...                               ...       ...          ...         ...
Wes Craven                       1984      2011           10          27
Wolfgang Petersen                1981      2006            7          25
Woody Allen                      1977      2013           18          36
Zack Snyder                      2004      2016            7          12
Zhang Yimou                      2002      2014            6          12

[199 rows x 4 columns]

***2) Then calculate rate of directing movies by title_count/yrs_active***

In [76]:
data_agg['movie_per_yr'] = data_agg['title_count']/data_agg['yrs_active']
data_agg

year_min  year_max  title_count  yrs_active  \
director_name                                                              
Adam McKay                       2004      2015            6          11   
Adam Shankman                    2001      2012            8          11   
Alejandro González Iñárritu      2000      2015            6          15   
Alex Proyas                      1994      2016            5          22   
Alexander Payne                  1999      2013            5          14   
...                               ...       ...          ...         ...   
Wes Craven                       1984      2011           10          27   
Wolfgang Petersen                1981      2006            7          25   
Woody Allen                      1977      2013           18          36   
Zack Snyder                      2004      2016            7          12   
Zhang Yimou                      2002      2014            6          12   

                             movie_per_yr  
director_name                              
Adam McKay                       0.545455  
Adam Shankman                    0.727273  
Alejandro González Iñárritu      0.400000  
Alex Proyas                      0.227273  
Alexander Payne                  0.357143  
...                                   ...  
Wes Craven                       0.370370  
Wolfgang Petersen                0.280000  
Woody Allen                      0.500000  
Zack Snyder                      0.583333  
Zhang Yimou                      0.500000  

[199 rows x 5 columns]

### Finally sorting this values

In [78]:
data_agg.sort_values('movie_per_yr',ascending=False)

year_min  year_max  title_count  yrs_active  movie_per_yr
director_name                                                              
Tyler Perry           2006      2013            9           7      1.285714
Jason Friedberg       2006      2010            5           4      1.250000
Shawn Levy            2002      2014           11          12      0.916667
Robert Rodriguez      1992      2014           16          22      0.727273
Adam Shankman         2001      2012            8          11      0.727273
...                    ...       ...          ...         ...           ...
Lawrence Kasdan       1985      2012            5          27      0.185185
Luc Besson            1985      2014            5          29      0.172414
Michael Apted         1980      2010            5          30      0.166667
Robert Redford        1980      2010            5          30      0.166667
Sidney Lumet          1976      2006            5          30      0.166667

[199 rows x 5 columns]